L2SIVacuum/POUs/Functions/Valves/FB_VCN_VAT590.TcPOU

```diff
@@ -278,7 +441,168 @@ stVcnVat590.stValveStatus.bPositionSPOutOfRange := i_stExtnWarnings.Bit6Position
 stVcnVat590.stValveStatus.bCtrlSPOutOfRange := i_stExtnWarnings.Bit10CtrlSPOutOfRange;
 stVcnVat590.stValveStatus.bGeneralCtrlSPOutOfRange := i_stExtnWarnings.Bit11GeneralCtrlSPOutOfRange;
 stVcnVat590.stValveStatus.bProcDataSettingsNotvalid := i_stExtnWarnings.Bit12ProcDataSettingsNotvalid;
-]]></ST>
+//Mapping WcState - Valve Ethercat Status
+stVcnVat590.bValveEthCatStatus := i_bValveEthCatStatus;]]></ST>
+      </Implementation>
+    </Action>
+    <Action Name="SetCoeInterface" Id="{8fba1572-d062-4a30-8a85-24af9871d9c6}">
+      <Implementation>
+        <ST><![CDATA[stVat590DriveRef.sNetId := F_CreateAmsNetId(i_stVat590PlcDriveRef.aNetId);
+stVat590DriveRef.nSlaveAddr := i_stVat590PlcDriveRef.nSlaveAddr;
+stVat590DriveRef.nDriveNo := i_stVat590PlcDriveRef.nDriveNo;
+stVat590DriveRef.nDriveType := i_stVat590PlcDriveRef.nDriveType;
+fbCoeRead.stDriveRef := stVat590DriveRef;
+fbCoeWrite.stDriveRef := stVat590DriveRef;]]></ST>
+      </Implementation>
+    </Action>
+    <Method Name="SetThrottle" Id="{bc87ba94-efcb-4d7b-ace0-cb07482f478f}">
+      <Declaration><![CDATA[METHOD SetThrottle : BOOL
+VAR_INPUT
+	rUpperLimit	:	REAL;
+END_VAR
+]]></Declaration>
+      <Implementation>
+        <ST><![CDATA[stVcnVat590.fPosSetLimit := LIMIT(0, rUpperLimit, 100);]]></ST>
+      </Implementation>
+    </Method>
+    <Method Name="ValveControl" Id="{25066385-eb5f-47fb-8ac4-a5fcdc50740f}">
+      <Declaration><![CDATA[METHOD ValveControl : BOOL
+VAR_INPUT
+	state : E_VCN_VAT590; // Set Control modes
+END_VAR
+]]></Declaration>
+      <Implementation>
+        <ST><![CDATA[stVcnVat590.eValveControl := state;]]></ST>
+      </Implementation>
+    </Method>
+    <Action Name="WriteCoeParameters" Id="{9f7f239a-46a5-4609-b528-6b15a6fa6552}">
+      <Implementation>
+        <ST><![CDATA[// User will change one parameter at a time so solution with ELSIF should work OK
```

Comment:
Even if the user changes more than one at a time this FB will still work as written, just doing one write at a time. Seems solid.

Reply:
Problem here was that in the ReadCoEInterface I overwrite internal structure containing CoE parameters with stVcnVat590.stCoeParameters structure which is IO to the EPICS interface every cycle. If user would change more then one parameter at a time the function will detect only one change because in the next cycle internal structure and stVcnVat590.stCoeParameters will be identical.

Reply:
Yes, that's true!